internal/hcs/system.go

```diff
@@ -326,13 +337,122 @@ func (computeSystem *System) Properties(ctx context.Context, types ...schema1.Pr
+	// process running on the machine. They then grab the pids that are running in the container and filter down
+	// the entries in the array to only what's running in that silo and start tallying up the total. This doesn't
+	// work well as performance should get worse if more processess are running on the machine in general and not
+	// just in the container. All of the additional information besides the WorkingSetPrivateSize field is ignored
+	// as well which isn't great and is wasted work to fetch.
+	//
+	// HCS only let's you grab statistics in an all or nothing fashion, so we can't just grab the private
+	// working set ourselves and ask for everything else seperately. The optimization we can make here is
+	// to open the silo ourselves and do the same queries for the rest of the info, as well as calculating
+	// the private working set in a more efficient manner by:
+	//
+	// 1. Find the pids running in the silo
+	// 2. Get a process handle for every process (only need PROCESS_QUERY_LIMITED_INFORMATION access)
+	// 3. Call NtQueryInformationProcess on each process with the class ProcessVmCounters
+	// 4. Tally up the total using the field PrivateWorkingSetSize in VM_COUNTERS_EX2.
+	privateWorkingSet, err := job.QueryPrivateWorkingSet()
+	if err != nil {
+		return nil, err
+	}
+
+	return &hcsschema.Statistics{
+		Timestamp:          timestamp,
+		ContainerStartTime: computeSystem.startTime,
+		Uptime100ns:        uint64(time.Since(computeSystem.startTime).Nanoseconds()) / 100,
+		Memory: &hcsschema.MemoryStats{
+			MemoryUsageCommitBytes:            memInfo.JobMemory,
+			MemoryUsageCommitPeakBytes:        memInfo.PeakJobMemoryUsed,
+			MemoryUsagePrivateWorkingSetBytes: privateWorkingSet,
+		},
+		Processor: &hcsschema.ProcessorStats{
+			RuntimeKernel100ns: uint64(processorInfo.TotalKernelTime),
+			RuntimeUser100ns:   uint64(processorInfo.TotalUserTime),
+			TotalRuntime100ns:  uint64(processorInfo.TotalKernelTime + processorInfo.TotalUserTime),
+		},
+		Storage: &hcsschema.StorageStats{
+			ReadCountNormalized:  uint64(storageInfo.ReadStats.IoCount),
+			ReadSizeBytes:        storageInfo.ReadStats.TotalSize,
+			WriteCountNormalized: uint64(storageInfo.WriteStats.IoCount),
+			WriteSizeBytes:       storageInfo.WriteStats.TotalSize,
+		},
+	}, nil
+}
+
 // PropertiesV2 returns the requested container properties targeting a V2 schema container.
-func (computeSystem *System) PropertiesV2(ctx context.Context, types ...hcsschema.PropertyType) (*hcsschema.Properties, error) {
+func (computeSystem *System) PropertiesV2(ctx context.Context, types ...hcsschema.PropertyType) (_ *hcsschema.Properties, err error) {
 	computeSystem.handleLock.RLock()
 	defer computeSystem.handleLock.RUnlock()
 
 	operation := "hcs::System::PropertiesV2"
 
+	// Check if any of the queries are for stats. We can grab these in process and skip the hop to
+	// HCS.
+	var statsPresent bool
```

Comment:
This current pattern seems like it will be difficult to extend to support other property types that we may want to query directly in the future. It would be nice if adding a new property type here was easier.

Reply:
That's what I was hoping with this https://github.com/microsoft/hcsshim/pull/1362#discussion_r866408248. This PR has devolved so much that swapping to that in the future (follow up PR) seemed easier

Reply:
One option is something like the below, although then we need to add a `case` for each PT that we might get from HCS.

```golang
if computeSystem.typ != "container" {
    return hcsPropertyQuery(ctx, computeSystem.handle, types)
}

properties := &hcsschema.Properties{
    Id: computeSystem.id,
    SystemType: computeSystem.typ,
    RuntimeOsType: computeSystem.os,
    Owner: computeSystem.Owner,
}
var fallbackQueryTypes []hcsschema.PropertyType
for _, propType := range types {
    switch propType {
    case hcsschema.PTStatistics:
        if properties.Statistics == nil {
            properties.Statistics, err = computeSystem.statisticsInProc(ctx)
            if err != nil {
                // error!
               fallbackQueryTypes = append(fallbackQueryTypes, propType)
            }
        }
    default:
        fallbackQueryTypes = append(fallbackQueryTypes, propType)
}

hcsProperties, err := hcsPropertyQuery(ctx, computeSystem.handle, fallbackQueryTypes)
if err != nil {
    // error!
}
for _, propType := range fallbackQueryTypes {
    switch propType {
    case hcsschema.PTStatistics:
        properties.Statistics = hcsProperties.Statistics
    }
    // TODO: Extend for other PTs.
}
```

Reply:
We could do that and simply leave everything but Statistics as automatically getting added to fallbacks, is this what you'd prefer? I already have a branch with this already essentially

Reply:
> That's what I was hoping with this [#1362 (comment)](https://github.com/microsoft/hcsshim/pull/1362#discussion_r866408248). This PR has devolved so much that swapping to that in the future (follow up PR) seemed easier

I had missed your reply on the other comment thread. I think we shouldn't leave things in a hacky state after any PRs. If there's cleanup to this code that should be done, I'm okay breaking it out into a separate PR, but then can we get that in before merging this one? That way everything stays clean.

Reply:
I'll have to think about what divide makes sense.. probably just end up using this PR.